python/pylibraft/pylibraft/cluster/kmeans.pyx

```diff
@@ -202,24 +202,33 @@ def compute_new_centroids(X,
 
 @auto_sync_handle
 @auto_convert_output
-def init_plus_plus(X, n_clusters, seed=None, handle=None):
+def init_plus_plus(X, n_clusters=None, seed=None, handle=None, centroids=None):
+    if n_clusters is not None and centroids is not None:
```

Comment:
Rather than failing if they are both specified, could we just fail here when `n_clusters != len(centroids)`? That way as long they match, it's okay to specify both.

Reply:
I'd be -1 on that. For me a caller that passes in both is signaling that they made a mistake or are confused (A bit like indicating left while being in a right turn lane with your car). The other thing I can't work out is when someone would use this possibility?

Reply:
From a user experience perspective, I'd prefer to try to handle any accidental mistakes by providing the expected behavior (when we know for sure their intention) without raising an error when possible.

The `centroids` argument is specifically to allow users to pre-allocate their output buffer up front instead of having us do the allocation inside this function. Yes, we can infer the number of clusters from this output buffer, but if specified, the output buffer is the source of truth for the number of centroids.

What I'm suggesting is that in the case where they've done "a little extra" by specifying both centroids argument and n_clusters, there's no harm done if n_clusters does actually match the number of centroids they've passed in. All they've done is provided some redundant information. In this case I can't see a reason for failing outright when the information they've provided is still correct.

Now, if they've specified both arguments and n_clusters doesn't match the number of centroids, then that's a different story. In that case the entropy is high- they've clearly gotten confused or done something wrong inadvertently.